src/modules/auth/auth.service.ts

```diff
@@ -0,0 +1,130 @@
+import * as jwt from 'jsonwebtoken';
+import { SigninDto } from '../user/dtos/signin.dto';
+import { UserResponseDTO } from '../user/dtos/response.dto';
+
+@Injectable()
+export class AuthService {
+  constructor(private prismaService: PrismaService) {}
+
+  async signup({
+    firstName,
+    lastName,
+    email,
+    password,
+    username,
+  }: SignupDto): Promise<{ token: string; user: UserResponseDTO }> {
+    const user = await this.prismaService.users.findUnique({
+      where: { email },
+    });
+
+    if (user) {
+      throw new ConflictException();
+    }
+
+    const hash = await bcrypt.hash(
+      password,
+      Number(process.env.BCRYPT_SALT_ROUNDS),
+    );
+
+    const newUser: Users = await this.prismaService.users.create({
+      data: {
+        firstName,
+        lastName,
+        email,
+        password: hash,
+        username,
+      },
+    });
+    delete newUser.password;
+    const token = await this.generateToken({ username, id: newUser.id });
+    return { token, user: { ...newUser } };
+  }
+  async oauth(body: any): Promise<{ token: string; user: UserResponseDTO }> {
+    const { email, firstName, lastName, googleId } = body;
+    const user = await this.prismaService.users.findUnique({
+      where: { googleId },
+    });
+
+    if (!user) {
+      let isUsernameTaken = true;
+      let generatedUsername = '';
+      while (isUsernameTaken) {
+        generatedUsername = await this.generateUsername(firstName, lastName);
+        isUsernameTaken = await this.checkUsername(generatedUsername);
+      }
+
+      const newUser: Users = await this.prismaService.users.create({
+        data: {
+          firstName,
+          lastName,
+          email,
+          googleId,
+          username: generatedUsername,
+        },
+      });
+      const token = await this.generateToken({
+        username: newUser.username,
+        id: newUser.id,
+      });
+      delete newUser.password;
+      return { token, user: newUser };
+    }
+
+    const token = await this.generateToken({
+      username: user.username,
+      id: user.id,
+    });
+    delete user.password;
+    return { token, user };
+  }
+
+  async signin({
+    email,
+    password,
+  }: SigninDto): Promise<{ token: string; user: UserResponseDTO }> {
+    const user = await this.prismaService.users.findUnique({
+      where: { email },
+    });
+
+    if (!user) {
+      throw new HttpException('Invalid Credentials', 400);
+    }
+
+    const hashedPassword = user.password;
+    const isValidPassword = await bcrypt.compare(password, hashedPassword);
+
+    if (!isValidPassword) {
+      throw new HttpException('Invalid Credentials', 400);
+    }
+
+    const token = await this.generateToken({
+      username: user.username,
+      id: user.id,
+    });
+    delete user.password;
+    return { token, user };
+  }
+
+  async generateToken({ username, id }) {
+    return jwt.sign({ username, id }, process.env.JWT_SECRET, {
+      expiresIn: '5 days',
```

Comment:
How the app would be more secure token expiration time should be 10 minutes, because tokens can be stolen. Maybe here would be good to put "TODO" comment that a refresh token will be implemented. The refresh token will be stored in the database and it will logout the user with the compromised token.